### PR TITLE
Add NorMuon optimizer and Cautious Weight Decay

### LINKED ISSUE #464: Two new optimization techniques: NorMuon and Cautious Weight Decay
NorMuon and Cautious Weight Decay have recently set two new records on the GPT speedrun repository and proven effective in their respective accompanying science papers. (with the latter technique exhaustively tested with 20,000 GPU hours)

Both methods work out-of-the-box, with minimal modifications to the Muon/GenericOptim class.

NorMuon: https://arxiv.org/pdf/2510.05491, https://github.com/zichongli5/NorMuon.

Cautious Weight Decay: https://arxiv.org/pdf/2510.12402, https://github.com/kozistr/pytorch_optimizer/commit/99dc4a536181892e4a8f2f60c292345d6dc1b566.

I think they would be much useful for training acceleration / convergence.

### PR DESCRIPTION
**Pull request by kabachuha**
NorMuon and Cautious Weight Decay have recently set two new records on the GPT speedrun repository and proven effective in their respective accompanying science papers. (with the latter technique exhaustively tested with 20,000 GPU hours)

Both methods work out-of-the-box, with minimal modifications to the Muon/GenericOptim class.

Cautious Weight Decay is intended to improve the convergence end result (more stable region), while NorMuon simply enables to reach the given loss faster.

In my experience, NorMuon works *slightly* slower than Muon *per step*, but if it achieves better loss, then it's acceptable.

Following ModdedGPT's code, I also added a "schedule" for weight decay - this way it is proportional to the learning rate, meaning it will gradually increase at the beginning of training and then gradually decrease in the end (assuming, learning rate schedule is enabled)

NorMuon: https://arxiv.org/pdf/2510.05491, https://github.com/zichongli5/NorMuon.

Cautious Weight Decay: https://arxiv.org/pdf/2510.12402, https://github.com/kozistr/pytorch_optimizer/commit/99dc4a536181892e4a8f2f60c292345d6dc1b566, https://github.com/KellerJordan/modded-nanogpt/pull/154/.

Closes #464.

### CODE REVIEW

File: optimizers/generic_optim.py [hunk 1]
```diff
@@ -555,3 +615,6 @@ def load_state_dict(self, sd):
                 # The projector is a class which contains tensors, so state needs to be explicitly moved to the correct device.
                 if 'projector' in state:
                     state['projector'].to(p.device)
+                # Handle second_momentum_buffer for NorMuon
```
Comment (tdrussell) [inline optimizers/generic_optim.py]: You don't need this, all state is already handled above.

File: optimizers/generic_optim.py [hunk 2]
```diff
@@ -353,6 +360,38 @@ def step(self, closure: Callable = None):
...
+                    # Update second momentum
+                    second_momentum.lerp_(v_mean, 1 - beta2)
+                    
+                    # Compute step size per element
+                    step_size_per_element = 1 / second_momentum.sqrt().add_(group["eps"])
+                    
+                    # Apply normalization
+                    numerator.mul_(step_size_per_element)
+                    
+                    # Compute new norm and rescale to maintain original norm
```
Comment (tdrussell) [inline optimizers/generic_optim.py]: What is this? I don't see it in the paper.

File: optimizers/generic_optim.py [hunk 3]
```diff
@@ -353,6 +360,38 @@ def step(self, closure: Callable = None):
...
+                    vnorm_new = numerator.norm(dim=(-2, -1), keepdim=True)
+                    numerator.mul_(vnorm / (vnorm_new + group["eps"]))
+                    
+                    # Apply the same scaling as Muon
+                    step_size *= math.sqrt(max(rows, cols))
```
Comment (tdrussell) [inline optimizers/generic_optim.py]: Should probably just directly implement Algorithm 1, line 10 in the paper, instead of this. They are very similar. For an orthonormal matrix, the frobenius norm is sqrt(min(rows, cols)). So if the update were exactly orthogonalized, that line in the paper reduces to this code. But it is approximately orthogonalized, not exact, so there is a difference.

OG Muon does it like this code, but better to just exactly stick with what the NorMuon paper says.

**Comment by kabachuha** [comment]
Okay, thank you for the review! I'll look into simplifying the NorMuon part.

To be transparent, this was vibe-coded based on Muon's, Normuon's, and diffusion-pipe codebases. 😅

The integration of CWD is fully organic, however

**Comment by tdrussell** [comment]
You can also just leave this PR open. I will want to test out NorMuon myself at some point, and I can make the changes myself when I do that if it's not done already.
